sklearn/tree/tree.py

```diff
@@ -238,10 +254,15 @@ def fit(self, X, y, sample_weight=None, check_input=True,
         if len(y) != n_samples:
             raise ValueError("Number of labels=%d does not match "
                              "number of samples=%d" % (len(y), n_samples))
-        if self.min_samples_split <= 0:
-            raise ValueError("min_samples_split must be greater than zero.")
-        if self.min_samples_leaf <= 0:
-            raise ValueError("min_samples_leaf must be greater than zero.")
+        if not (0. < self.min_samples_split <= 1. or
+                2 <= self.min_samples_split):
+            raise ValueError("min_samples_split must be in at least 2"
+                             " or in (0, 1], got %s" % min_samples_split)
+        if not (0. < self.min_samples_leaf <= 0.5 or
+                1 <= self.min_samples_leaf):
+            raise ValueError("min_samples_leaf must be at least than 1 "
```

Comment:
I have added the upperbound `self.min_samples_leaf < 0.5 * n_samples ` as this currently works without any problems. You just have a pre-pruning constraint that won't let you grow the tree.

If we add this upper-bound, this could break user code and be painful while performing grid-search.